### PR TITLE
(PUP-10163) Add thread safety for singletons using Visitor

### PR DESCRIPTION
**Pull request by pcarlisle**
Several singletons use Puppet::Pops::Visitor which has some caching. To
be sure that this is safe, implement these using ThreadLocalSingleton.
Some of these used class variables referred to from inside the singleton
instances. This seems redundant and should be safe to make instance
variables.

### CODE REVIEW

File: lib/puppet/pops/model/pn_transformer.rb [hunk 1]
```diff
@@ -1,22 +1,18 @@
...
-  end
+  extend Puppet::Concurrent::ThreadLocalSingleton
 
   def self.transform(ast)
     singleton.transform(ast)
   end
 
   def initialize(visitor)
-    @visitor = visitor
+    @visitor = Visitor.new(nil, 'transform', 0, 0)
```
Comment (mwaggett) [inline lib/puppet/pops/model/pn_transformer.rb]: Are we intentionally no longer using the `visitor` that gets passed in?

Comment (pcarlisle) [reply lib/puppet/pops/model/pn_transformer.rb]: good catch, i removed that

Comment (mwaggett) [reply lib/puppet/pops/model/pn_transformer.rb]: Are there things calling this that will need to be updated? I'm not familiar with how this is used at all.

Comment (Magisus) [reply lib/puppet/pops/model/pn_transformer.rb]: I think it's just the creation of the singleton, since we try to use that everywhere.

**Comment by Magisus** [comment]
Jenkins please test this with servertests

**Comment by justinstoller** [comment]
I haven't seen branches for the latest releases so I assume we want to wait to merge this until after that happens?

**Comment by Magisus** [comment]
There are release branches for puppet-agent; they don't usually cut them for Puppet. So I think you can go ahead if you think this is ready.

**Pull request merged**